### PR TITLE
Issue 5725 spine extension apis

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: build_tools/waf_dynamo.py [hunk 1]
```diff
@@ -176,6 +176,9 @@ def extract_docs(bld, src):
         # Gather data
         for s in src:
             n = bld.path.find_resource(s)
+            if not n:
+                print("Couldn't find resource: %s" % s)
```
Comment (JCash) [inline build_tools/waf_dynamo.py:180]: Better with a printout than an unreadable callstack

File: com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/textureset/test/TextureSetGeneratorTest.java [hunk 2]
```diff
@@ -29,9 +29,9 @@
 import com.dynamo.bob.textureset.TextureSetGenerator.AnimIterator;
 import com.dynamo.bob.textureset.TextureSetGenerator.TextureSetResult;
 import com.dynamo.bob.textureset.TextureSetGenerator.UVTransform;
-import com.dynamo.textureset.proto.TextureSetProto.TextureSet;
-import com.dynamo.textureset.proto.TextureSetProto.TextureSetAnimation;
-import com.dynamo.tile.proto.Tile.Playback;
+import com.dynamo.gamesys.proto.TextureSetProto.TextureSet;
+import com.dynamo.gamesys.proto.TextureSetProto.TextureSetAnimation;
+import com.dynamo.gamesys.proto.Tile.Playback;
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob.test/src/com/dynamo/bob/textureset/test/TextureSetGeneratorTest.java:34]: Unifying the namespaces a bit, so it's easier to know where to find the source code

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/GameObjectBuilder.java [hunk 3]
```diff
@@ -197,6 +196,12 @@ public void build(Task<Void> task) throws CompileExceptionError,
             for (String[] fromTo : extensionMapping) {
                 c = BuilderUtil.replaceExt(c, fromTo[0], fromTo[1]);
             }
+
+            // Use the BuilderParams from each builder to map the input ext to the output ext
+            String inExt = "." + FilenameUtils.getExtension(c);
+            String outExt = project.replaceExt(inExt);
+            c = BuilderUtil.replaceExt(c, inExt, outExt);
```
Comment (JCash) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/GameObjectBuilder.java:203]: Makes the lookup data driven, by using the BuilderParams class

File: engine/dlib/src/dlib/hash.h [hunk 4]
```diff
@@ -23,122 +23,6 @@ const uint32_t DMHASH_MAX_REVERSE_LENGTH = 1024U;
 extern "C"
 {
 
-/**
```
Comment (JCash) [inline engine/dlib/src/dlib/hash.h:26]: Moved to dmsdk

File: engine/engine/src/engine.cpp [hunk 5]
```diff
@@ -998,10 +996,6 @@ namespace dmEngine
         engine->m_MeshContext.m_Factory       = engine->m_Factory;
         engine->m_MeshContext.m_MaxMeshCount = dmConfigFile::GetInt(engine->m_Config, "mesh.max_count", 128);
 
-        engine->m_SpineModelContext.m_RenderContext = engine->m_RenderContext;
-        engine->m_SpineModelContext.m_Factory = engine->m_Factory;
-        engine->m_SpineModelContext.m_MaxSpineModelCount = max_spine_count;
-
```
Comment (JCash) [inline engine/engine/src/engine.cpp:1004]: Another step towards removal, making the default implementation into an actual extension.

File: engine/gameobject/src/dmsdk/gameobject/res_lua.h [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+#ifndef DMSDK_GAMEOBJECT_RES_LUA_H
+#define DMSDK_GAMEOBJECT_RES_LUA_H
+
+#include <stdint.h>
+
+#include <ddf/ddf.h>
+#include <gameobject/lua_ddf.h>
```
Comment (JCash) [inline engine/gameobject/src/dmsdk/gameobject/res_lua.h:19]: The "_ddf.h" aren't under the actual "dmsdk" folder, in part to signal that they are more "volatile" than the other headers.

File: engine/gamesys/src/dmsdk/gamesys/resources/res_collision_object.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+        dmPhysics::HCollisionShape3D* m_Shapes3D;
+        dmPhysics::HCollisionShape2D* m_Shapes2D;
+        dmVMath::Vector3* m_ShapeTranslation;
+        dmVMath::Quat* m_ShapeRotation;
+        uint32_t m_TileGridShapeCount;
+        uint32_t m_ShapeCount;
+
+        dmPhysicsDDF::CollisionObjectDesc* m_DDF;
+        uint32_t m_TileGrid : 1;
```
Comment (Jhonnyg) [inline engine/gamesys/src/dmsdk/gamesys/resources/res_collision_object.h:47]: Can we use a bit from either m_TileGridShapeCount or m_ShapeCount?

File: engine/gamesys/src/gamesys/scripts/script_label.cpp [hunk 8]
```diff
@@ -204,6 +199,12 @@ namespace dmGameSystem
  * end
  * ```
  */
+
+
+    // As seen in gamesys_private.h (which makes it a _lot_ harder to search for)
+    static const char* LABEL_EXT = "labelc";
+    //static const dmhash_t LABEL_EXT_HASH = dmHashString64(LABEL_EXT);
```
Comment (Jhonnyg) [inline engine/gamesys/src/gamesys/scripts/script_label.cpp:206]: junk?

File: engine/graphics/src/graphics.h [hunk 9]
```diff
@@ -588,28 +479,13 @@ namespace dmGraphics
      */
     void Clear(HContext context, uint32_t flags, uint8_t red, uint8_t green, uint8_t blue, uint8_t alpha, float depth, uint32_t stencil);
 
-    HVertexBuffer NewVertexBuffer(HContext context, uint32_t size, const void* data, BufferUsage buffer_usage);
-    void DeleteVertexBuffer(HVertexBuffer buffer);
-    void SetVertexBufferData(HVertexBuffer buffer, uint32_t size, const void* data, BufferUsage buffer_usage);
-    void SetVertexBufferSubData(HVertexBuffer buffer, uint32_t offset, uint32_t size, const void* data);
+    // Are these used somewhere?
```
Comment (Jhonnyg) [inline engine/graphics/src/graphics.h:482]: Only in the tests afaik, map / unmap isn't really used in Opengl anymore

**Review by Jhonnyg — DISMISSED** [review]
Awesome! Alot to digest, but it looks good in my book :)

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ProtoBuilder.java [hunk 10]
```diff
@@ -44,6 +44,11 @@ static public void addMessageClass(String ext, Class<? extends GeneratedMessage>
         return extToMessageClass.get(ext);
     }
 
+    static public boolean supportsType(String ext) {
+        Class<? extends GeneratedMessage> klass = getMessageClassFromExt(ext);
+        return klass != null;
```
[the remark this answers is not in the document]
Comment (britzl) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ProtoBuilder.java:49]: I believe this is to not confuse things with a reserved `Class` keyword in Java. "klass" or "clazz" is not uncommon in code like this.

I think I'd prefer to just name the variable "c" or "cls" though :-)

**Review by britzl — DISMISSED** [review]
I've only briefly skimmed through the changes. Amazing work!

File: engine/gamesys/src/gamesys/scripts/script_label.cpp (hunk 8 above)
Comment (JCash) [reply engine/gamesys/src/gamesys/scripts/script_label.cpp:206]: well, I left it as "documentation", but perhaps I should just remove it

**Pull request merged**